File: Koko-Eating-Bananas.cs
// Binary Search -- O(n * log(max(piles)))
public class Solution {
    public int MinEatingSpeed(int[] piles, int h) {
        long low = 1;
        long high = piles.Max();

        while (low <= high) {
            long k = (low + high) / 2;
            long hours = 0;

            foreach (int pile in piles) {
                hours += (long)Math.Ceiling((double)pile / k);
                if (hours > h) {
                    break; 
                }
            }

            if (hours <= h) {
                high = k - 1;
            } else {
                low = k + 1;
            }
        }

        return (int)low;
    }
}

File: Koko-Eating-Bananas.cpp
// Binary Search
// O(Len(piles)*Log(Max(piles)))
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        long long low = 1;
        auto max_it = max_element(piles.begin(), piles.end());
        long long high = *max_it;
        while(low<=high){
            long long k = low + (high - low) / 2; 
            long long hours = 0 ;
            for(int pile : piles){
                hours += (static_cast<long long>(pile) + k - 1) / k;
            }
            if(hours>h){
                low = k + 1 ;
            }else{
                high = k-1;
            }

        }

        return static_cast<int>(low);
    }
};